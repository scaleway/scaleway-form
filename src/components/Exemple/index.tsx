import React from 'react'

const Exemple = (): JSX.Element => <div>This is a test</div>

export default Exemple
