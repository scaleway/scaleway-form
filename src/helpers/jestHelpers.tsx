import { ThemeProvider } from '@emotion/react'
import makeHelpers from '@scaleway/jest-helpers'
import { theme as scwuiTheme } from '@scaleway/ui'
import { RenderResult, render } from '@testing-library/react'
import React, { FC, ReactElement, ReactNode } from 'react'
import Form from '../components/Form'
import mockErrors from '../mocks/mockErrors'

interface WrapperProps {
  theme?: typeof scwuiTheme
}

const Wrapper: FC<WrapperProps> = ({ theme = scwuiTheme, children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotWithPortal,
} = makeHelpers(Wrapper, { classNameReplacer: className => className })

export const shouldMatchEmotionSnapshotFormWrapper = (
  children: ReactNode,
  options?: {
    transform: (node: RenderResult) => Promise<void> | void
  },
) =>
  shouldMatchEmotionSnapshot(
    <Form errors={mockErrors}>{() => children}</Form>,
    options,
  )

export const renderWithWrapper = (children: ReactElement) =>
  render(children, {
    wrapper: () => (
      <ThemeProvider theme={scwuiTheme}>
        <Form errors={mockErrors}>{children}</Form>
      </ThemeProvider>
    ),
  })

export const mockRandom = () =>
  jest.spyOn(global.Math, 'random').mockReturnValue(0.4155913669444804)

export const restoreRandom = () =>
  jest.spyOn(global.Math, 'random').mockRestore()
