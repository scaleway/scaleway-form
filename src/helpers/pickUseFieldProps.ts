import { AnyObject } from 'final-form'
import { UseFieldConfig } from 'react-final-form'
import { BaseFieldProps } from '../types'

const pickUseFieldProps = <T = AnyObject, K = AnyObject>({
  afterSubmit,
  allowNull,
  beforeSubmit,
  data,
  defaultValue,
  format,
  formatOnBlur,
  initialValue,
  isEqual,
  multiple,
  parse,
  subscription,
  validate,
  validateFields,
  value,
}: BaseFieldProps<T, K>): UseFieldConfig<T, K> => ({
  afterSubmit,
  allowNull,
  beforeSubmit,
  data,
  defaultValue,
  format,
  formatOnBlur,
  initialValue,
  isEqual,
  multiple,
  parse,
  subscription,
  validate,
  validateFields,
  value,
})

export default pickUseFieldProps
