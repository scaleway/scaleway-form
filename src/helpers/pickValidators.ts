import type { ValidatorObject, ValidatorProps } from '../types'
import validators from '../validators'
import type { ValidatorFn } from '../validators/types'

export const pickValidators = <InputValue = unknown>(args: ValidatorProps) =>
  Object.entries(args)
    .map(([key, value]) =>
      value !== undefined
        ? (validators as Record<string, ValidatorFn>)[
            key as keyof typeof validators
          ]?.(value)
        : undefined,
    )
    .filter(validator => !!validator) as ValidatorObject<InputValue>[]
