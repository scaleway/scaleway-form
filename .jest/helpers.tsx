import { ThemeProvider } from '@emotion/react'
import makeHelpers from '@scaleway/jest-helpers'
import { theme as lightTheme } from '@scaleway/ui'
import { render } from '@testing-library/react'
import type { ComponentProps, ReactElement, ReactNode } from 'react'
import { Form } from '../src/components'
import { mockErrors } from '../src/mocks'

type WrapperProps = {
  theme?: typeof lightTheme
  children: ReactNode
}

const Wrapper = ({ theme = lightTheme, children }: WrapperProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotWithPortal,
} = makeHelpers(Wrapper, { classNameReplacer: className => className })

export const shouldMatchEmotionSnapshotFormWrapper = (
  children: ReactNode,
  options?: Parameters<typeof shouldMatchEmotionSnapshot>[1],
  formOptions?: Partial<ComponentProps<typeof Form>>,
) => {
  const { initialValues } = formOptions ?? {}

  return shouldMatchEmotionSnapshot(
    <Form
      onRawSubmit={() => {}}
      errors={mockErrors}
      initialValues={initialValues}
    >
      {() => children}
    </Form>,
    options,
  )
}

export const renderWithWrapper = (children: ReactElement) =>
  render(children, {
    wrapper: () => (
      <ThemeProvider theme={lightTheme}>
        <Form onRawSubmit={() => {}} errors={mockErrors}>
          {children}
        </Form>
      </ThemeProvider>
    ),
  })

export const mockRandom = () =>
  jest.spyOn(global.Math, 'random').mockReturnValue(0.4155913669444804)

export const restoreRandom = () =>
  jest.spyOn(global.Math, 'random').mockRestore()
