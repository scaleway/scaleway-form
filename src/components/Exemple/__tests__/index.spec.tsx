import React from 'react'
import Exemple from '..'
import { shouldMatchEmotionSnapshot } from '../../../helpers/jestHelpers'

describe('Exemple', () => {
  test('renders correctly ', () => shouldMatchEmotionSnapshot(<Exemple />))
})
