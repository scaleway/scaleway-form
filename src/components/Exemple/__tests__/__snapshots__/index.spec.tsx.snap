// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exemple renders correctly  1`] = `
<DocumentFragment>
  <div>
    This is a test
  </div>
</DocumentFragment>
`;
